fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -2548,4 +2644,220 @@ public boolean isContinueIfWriteOnly() {
+        return getRunner().runAsync(context ->
+        {
+            RangeSet rangeSet = new RangeSet(subspace.subspace(Tuple.from(FDBRecordStore.INDEX_RANGE_SPACE_KEY, index.getSubspaceKey())));
+            TransactionContext tc = context.ensureActive();
+            return rangeSet.insertRange(tc, null, null).thenApply(bignore -> null);
+        });
+    }
+
+    private void buildIndexFromIndexThrowEx() {
+        throw new MetaDataException(IndexFromIndexRequest.BAD_SRC_MESSAGE);
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndexReadableOrThrow(FDBRecordStore store, String srcIndex) {
+        return store.getRecordStoreStateAsync()
+                .thenCompose(state ->
+                {
+                    if (!state.isReadable(srcIndex)) {
+                        buildIndexFromIndexThrowEx();
+                    }
+                    return AsyncUtil.DONE;
+                });
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndexAsync() {
+        // zz 0 - sanity & trigger
+        return getRunner().runAsync(this::openRecordStore)
+                .thenCompose(store ->
+                {
+                    // first verify that both src and tgt are of a single, similar, type
+                    final RecordMetaData metaData = store.getRecordMetaData();
+                    final Collection<RecordType> srcRecordTypes = metaData.recordTypesForIndex(index);
+
+                    if ( syntheticIndex ||
+                            recordTypes.size() != 1 ||
+                            srcRecordTypes.size() != 1 ||
+                            recordTypes.toArray()[0] != srcRecordTypes.toArray()[0]) {
+
+                        buildIndexFromIndexThrowEx();
+                    }
+
+                    // return the build iteration. src readability & tgt/src idempotency will be verified during each iteration
+                    return buildIndexFromIndex();
+                });
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndex() {
+        // zz 1 - runner
+        return getRunner().runAsync(context ->
+                context.getReadVersionAsync().thenCompose(vignore -> // Why do we need this version?
+                {
+                    SubspaceProvider subspaceProvider = recordStoreBuilder.getSubspaceProvider();
+                    assert subspaceProvider != null;
+                    // zz 2 - subspace
+                    return subspaceProvider.getSubspaceAsync(context)
+                            .thenCompose(subspace ->
+                                    buildIndexFromIndex(subspaceProvider, subspace)
+                                            .thenCompose(vignore2 -> buildIndexFromIndexMarkBuilt(subspace)
+                                            ));
+                }));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndex(SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
```

Comment:
Looks good generally but I didn't dig too deep into this method and the one below. It might be helpful to copy the code from the original path and modify it, when validating idea. But duplicated code make it impossible to find the differences (and possible issues) and keep the two path synchronized. I'll have another round of review when you finish refactoring.